visualization-project-laptop-sales/visualization_project_laptop_sales_solution.py

```diff
@@ -0,0 +1,126 @@
+"""
+
+# Let's see what happens if the the customer takes on the shipping cost
+sales['Sale Price Plus Shipping'] = sales['Sale Price'] + sales['Shipping Cost']
+consumer_cost2 = sales.groupby('Contact Sex')['Sale Price Plus Shipping'].sum()
+consumer_cost2
+
+# Now consider the profit percentage given customers take on the shipping cost
+# This gives the profit percentage per specific customer
+sales['Profit With Their Increased Cost'] = sales['Sale Price Plus Shipping'] - sales['Our Cost']
+sales['Our Profit Percentage With Their Increased Cost'] = (sales['Profit With Their Increased Cost'] / sales['Our Cost']) * 100
+sales['Our Profit Percentage With Their Increased Cost']
+
+# Net profit, grouped by sex
+net_profit2 = sales.groupby('Contact Sex')['Profit With Their Increased Cost'].sum()
+net_profit2
+
+# Net cost, grouped by sex
+net_cost2 = sales.groupby('Contact Sex')['Our Cost'].sum()
+net_cost2
+
+profit_percentage2 = net_profit2 / net_cost2
+profit_percentage2.rename('Profit Percentage', inplace = True)
+
+# Net profit percentage for all customers male and female
+total_profit_percentage1 = (sales['Profit With Their Increased Cost'].sum() / sales['Our Cost'].sum()) * 100
+total_profit_percentage1
+
+"""3. If the consumer is cash-constrained, males should be targeted.
+
+# **#4 What is our target age to maximize profit?**
+"""
+
+# First consider the individual ages
+
+individual_ages = sales.groupby('Contact Age')['Contact Age'].value_counts()
+individual_ages
+
+ages = pd.cut(np.array(list(range(20, 61, 10))),10)
+ages
+
+# Consider age ranges now
+age_ranges1 = sales.groupby('Contact Age')['Profit'].sum()
+
+sales['Age Ranges'] = pd.cut(sales['Contact Age'], bins = range(20, 61, 5), right = False)
+sales['Age Ranges']
+
+sales.groupby('Age Ranges')['Profit'].sum().plot.bar()
+
+"""4. Considering the invididual age counts and age ranges, the business might consider targeting customers who are between 45 (inclusive) to 50 years old (non-inclusive).
+
+# **#5 Which product should we feature?**
+"""
+
+# Consider 'Profit' grouped by 'Prduct Type'
+profit_by_prodtype = sales.groupby('Product Type')['Profit'].sum()
+profit_by_prodtype
+
+"""5. Since laptops yield the highest profit compared to desktops and tablets, the business should feature laptops.
+
+# **#6 What lead sources have worked in the past: website, flyer, or email?**
+"""
+
+# Consider 'Profit' grouped by Lead Source
+profit_by_leadsource = sales.groupby('Lead Source')['Profit'].sum()
+profit_by_leadsource
+
+"""6. The profit generated through Flyers 1, 2, 3, and 4 combined outyield profit genereated through email and website. However, if looking at the Flyers individually, then Website outyields the rest.
+
+# **#7 When is the best time to do email marketing?**
+"""
+
+# Consider profit grouped by lead source and then sale month
+
+profits_grouped = sales.groupby(['Lead Source', 'Sale Month'])['Profit'].sum()
```

Comment:
Your answer is correct, but an easier way to see it would be to filter out only rows that use Email and then group by month.